### PR TITLE
Add ForEachN from CUB to cccl/c.

### PR DESCRIPTION
**Pull request by wmaxey**
## Checklist

- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: c/src/for.cu [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+  if (num_items == 0)
+  {
+    return error;
+  }
+
+  small_aligned_storage<for_each_kernel_params> op_params =
+    (op.type == cccl_op_kind_t::stateful)
+      ? small_aligned_storage<for_each_kernel_params>({d_in.state}, op.state, op.size, op.alignment)
+      : small_aligned_storage<for_each_kernel_params>({d_in.state});
```
Comment (gevtushenko) [inline c/src/for.cu:71]: important: this is only valid for stateless iterators

File: c/src/util/small_storage.cpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#include <cstring>
+#include <memory>
+
+// copy_or_allocate_into_aligned_storage is a helper that will allocate space to hold an object and some aligned data
+// after it in one allocation. A return value of 0 indicates that the buffer pointed to by 'space' was sufficient for
+// holding the additional data. a non-zero return means a buffer was allocated large enough to include the
+// minimum size as determined by the offset.
+void* copy_or_allocate_into_aligned_storage(
```
Comment (gevtushenko) [inline c/src/util/small_storage.cpp:8]: suggestion: the large stateful op test is failing. Let's a standalone test for the small storage

File: c/src/util/small_storage.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  void* aligned_addr         = std::align(src_alignment, src_size, buffer_addr, space_size);
+  void* allocation           = nullptr;
+
+  if (!aligned_addr)
+  {
+    // Allocate enough space to fit the entirety of the buffer, this may overallocate by at most
+    // src_alignment - offset
+    size_t allocation_size = offset + src_alignment + src_size;
+
+    allocation           = malloc(allocation_size);
```
Comment (gevtushenko) [inline c/src/util/small_storage.cpp:28]: suggestion: consider memsetting this allocation in debug builds

File: c/src/for.cu [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+  cudaError error = cudaSuccess;
+
+  if (num_items == 0)
+  {
+    return error;
+  }
+
+  small_aligned_storage<for_each_kernel_params> op_params =
+    (op.type == cccl_op_kind_t::stateful)
+      ? small_aligned_storage<for_each_kernel_params>({d_in.state}, op.state, op.size, op.alignment)
```
Comment (gevtushenko) [inline c/src/for.cu:70]: critical: it's alignment and then size:
```suggestion
      ? small_aligned_storage<for_each_kernel_params>({d_in.state}, op.state, op.alignment, op.size)
```

**Review by gevtushenko — COMMENTED** [review]
This PR has most of the things we need already! Let's add iterators support and mark ready for review.

File: c/src/util/small_storage.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+template <typename T, size_t Extra = 8>
+struct small_aligned_storage
+{
+  union
+  {
+    small_aligned_storage_allocated<T> allocated;
+    small_aligned_storage_stack<T, Extra> local;
+  };
+
+  bool is_allocated = false;
```
[the remark this answers is not in the document]
Comment (wmaxey) [reply c/src/util/small_storage.h]: Yes, but I've removed the small vector class for now.

File: c/src/for.cu [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+  try
+  {
+    nvrtcProgram prog{};
+    const char* name = "test";
+
+    const int cc                     = cc_major * 10 + cc_minor;
+    const std::string d_data_value_t = cccl_type_enum_to_string(d_data.value_type.type);
+    const std::string offset_t       = cccl_type_enum_to_string(cccl_type_enum::UINT64);
+
+    const std::string for_kernel_name   = get_device_for_kernel_name("size_t");
```
[the remark this answers is not in the document]
Comment (wmaxey) [reply c/src/for.cu]: I believe this is fixed now.

File: c/src/util/small_storage.cpp (hunk 2 above)
Comment (wmaxey) [reply c/src/util/small_storage.cpp]: Will save unit tests for later.

**Comment by wmaxey** [comment]
pre-commit.ci autofix

File: c/test/test_for.cpp [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  const char* libcudacxx_path = TEST_LIBCUDACXX_PATH;
+  const char* ctk_path        = TEST_CTK_PATH;
+
+  cccl_device_for_build_result_t build;
+  REQUIRE(
+    CUDA_SUCCESS
+    == cccl_device_for_build(&build, input, op, cc_major, cc_minor, cub_path, thrust_path, libcudacxx_path, ctk_path));
+  const std::string sass = inspect_sass(build.cubin, build.cubin_size);
+  // REQUIRE(sass.find("LDL") == std::string::npos);
+  // REQUIRE(sass.find("STL") == std::string::npos);
```
Comment (gevtushenko) [inline c/test/test_for.cpp:37]: suggestion: let's return these checks:

```suggestion
  REQUIRE(sass.find("LDL") == std::string::npos);
  REQUIRE(sass.find("STL") == std::string::npos);
```

File: c/src/for.cu [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+  const char* cub_path,
+  const char* thrust_path,
+  const char* libcudacxx_path,
+  const char* ctk_path) noexcept
+{
+  CUresult error = CUDA_SUCCESS;
+
+  if (d_data.type == cccl_iterator_kind_t::iterator)
+  {
+    throw std::runtime_error(std::string("Iterators are unsupported in for_each currently"));
```
Comment (gevtushenko) [inline c/src/for.cu:84]: question: terminating the program in this case seems a bit extreme. Can we return an error instead?

File: c/src/for.cu [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+extern "C" CCCL_C_API CUresult cccl_device_for(
+  cccl_device_for_build_result_t build,
+  cccl_iterator_t d_data,
+  int64_t num_items,
+  cccl_op_t op,
+  CUstream stream) noexcept
+{
+  CUresult error = CUDA_SUCCESS;
+
```
Comment (gevtushenko) [inline c/src/for.cu:190]: suggestion: might be missing `pushed = try_push_context();`

**Review by gevtushenko — APPROVED** [review]

**Pull request merged**